camligorkem/assignment2/BinaryTree.java

```diff
@@ -0,0 +1,164 @@
+        if(root.getData().equals(key)){    // key found - previous node is its ancestor
+            ArrayList<T> ancestorList = new ArrayList<>();
+            return ancestorList;
+        }
+        ArrayList<T> ancestorsL = getAncestor(root.getLeft(), key);
+        ArrayList<T> ancestorsR = getAncestor(root.getRight(), key);
+        if(ancestorsL!= null || ancestorsR!=null){ // key is present in the tree
+            if(ancestorsL!=null){  // key is in the left subtree
+                ancestorsL.add(root.getData());
+                return ancestorsL;
+            }else{      // key is in the right subtree
+                ancestorsR.add(root.getData());
+                return ancestorsR;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Prints the ancestor list passed as parameter.
+     * If list is empty or null instead of list prints out a warning.
+     * @param list takes a ArrayList of generic type T
+     */
+    public void printAncestor(ArrayList<T> list){
+        if(list == null){
+            System.out.println("No ancestor found! Key doesn't exist");
+            return;
+        }
+        if(list.isEmpty()){
+            System.out.println("No ancestor found! Key is a root.");
+            return;
+        }
+        for(T element:list){
+            System.out.print(element+"\t");
+        }
+        System.out.println();
+    }
+
+    /**
+     * Prints the lowest common ancester of 2 keys
+     * Remark: if one key is ancestor of another case (i.e. 2,6 in assignment):
+     *     Ancestors of 2 = 3, 7 and ancestors of 6 = 2,3,7 and  common ancestors are 3,7;
+     *     Lowest common ancestor than should be 3.
+     * Remark 2: Since in getAncestor function when a duplicate exists, it returns the ancestor result of first one found,
+     * lowerCommonAncestor function for same values doesn't make sense to use for duplicates because it won't be between the duplicate
+     * nodes but the first found node with itself.
+     * @param key1
+     * @param key2
+     * @return the lowest common ancestor
+     */
+    public Optional<T> lowerCommonAncestor(T key1, T key2){
+        ArrayList<T> key1Ancestors = getAncestor(root,key1);
+        ArrayList<T> key2Ancestors = getAncestor(root,key2);
+        //no common ancestors -> one of the keys doesn't exist
+        if(key1Ancestors == null || key2Ancestors==null){
+            return Optional.empty();
+        }
+        // returns empty optional one of the keys is the root - no common ancestor since one has no ancestor
+        if(key1Ancestors.size() == 0 || key2Ancestors.size()==0)
+            return Optional.empty();
+        T lowestComAnc = null;
+        //reverse the lists and compare ancestor from key1 and key2 at the same index
+        // add them to commonAncestor list until you found one that doesn't match.
+        Collections.reverse(key1Ancestors);
+        Collections.reverse(key2Ancestors);
+        int size;
```

Comment:
This conditional assigning can be nicely replaced with:
int size = key1Ancestors.size() <= key2Ancestors.size() ? key1Ancestors.size() : key2Ancestors.size();

Reply:
I really like it. Thanks! Done!